cmd/chihaya/main.go

```diff
@@ -12,178 +11,138 @@ import (
-					defer pprof.StopCPUProfile()
-				}
-
-				configFile, err := ParseConfigFile(configFilePath)
-				if err != nil {
-					return errors.New("failed to read config: " + err.Error())
-				}
-				cfg := configFile.MainConfigBlock
-
-				go func() {
-					promServer := http.Server{
-						Addr:    cfg.PrometheusAddr,
-						Handler: prometheus.Handler(),
-					}
-					log.Println("started serving prometheus stats on", cfg.PrometheusAddr)
-					if err := promServer.ListenAndServe(); err != nil {
-						log.Fatal(err)
-					}
-				}()
-
-				// Force the compiler to enforce memory against the storage interface.
-				peerStore, err := memory.New(cfg.Storage)
-				if err != nil {
-					return err
-				}
-
-				// TODO create Hooks
-				logic := middleware.NewLogic(cfg.Config, peerStore, nil, nil, nil, nil)
-				if err != nil {
-					return err
-				}
-
-				shutdown := make(chan struct{})
-				errChan := make(chan error)
-
-				var httpFrontend *httpfrontend.Frontend
-				var udpFrontend *udpfrontend.Frontend
-
-				if cfg.HTTPConfig.Addr != "" {
-					httpFrontend = httpfrontend.NewFrontend(logic, cfg.HTTPConfig)
-
-					go func() {
-						log.Println("started serving HTTP on", cfg.HTTPConfig.Addr)
-						if err := httpFrontend.ListenAndServe(); err != nil {
-							errChan <- errors.New("failed to cleanly shutdown HTTP frontend: " + err.Error())
-						}
-					}()
-				}
-
-				if cfg.UDPConfig.Addr != "" {
-					udpFrontend = udpfrontend.NewFrontend(logic, cfg.UDPConfig)
-
-					go func() {
-						log.Println("started serving UDP on", cfg.UDPConfig.Addr)
-						if err := udpFrontend.ListenAndServe(); err != nil {
-							errChan <- errors.New("failed to cleanly shutdown UDP frontend: " + err.Error())
-						}
-					}()
-				}
-
-				sigChan := make(chan os.Signal)
-				signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
-				go func() {
-					select {
-					case <-sigChan:
-					case <-shutdown:
-					}
-
-					if udpFrontend != nil {
-						udpFrontend.Stop()
-					}
-
-					if httpFrontend != nil {
-						httpFrontend.Stop()
-					}
-
-					for err := range peerStore.Stop() {
-						if err != nil {
-							errChan <- err
-						}
-					}
-
-					close(errChan)
-				}()
-
-				closed := false
-				var bufErr error
-				for err = range errChan {
-					if err != nil {
-						if !closed {
-							close(shutdown)
-							closed = true
-						} else {
-							log.Println(bufErr)
-						}
-						bufErr = err
-					}
-				}
-
-				return bufErr
-			}(); err != nil {
+			if err := rootCmdRun(cmd, args); err != nil {
 				log.Fatal(err)
 			}
 		},
 	}
-
-	rootCmd.Flags().StringVar(&configFilePath, "config", "/etc/chihaya.yaml", "location of configuration file")
-	rootCmd.Flags().StringVarP(&cpuProfilePath, "cpuprofile", "", "", "location to save a CPU profile")
+	rootCmd.Flags().String("config", "/etc/chihaya.yaml", "location of configuration file")
+	rootCmd.Flags().StringP("cpuprofile", "", "", "location to save a CPU profile")
```

Comment:
I just noticed: You made this a string parameter with `""` as the shorthand - intended? Why?
Also, let's give `config` a shorthand, now that I know about the feature :)